etc/systemd/system-generators/zfs-mount-generator.c

```diff
@@ -0,0 +1,1056 @@
+/*
+ * Permission to use, copy, modify, and/or distribute this software for
+ * any purpose with or without fee is hereby granted.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
+ * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+
+#include <sys/resource.h>
+#include <sys/types.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <sys/mman.h>
+#include <semaphore.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <time.h>
+#include <regex.h>
+#include <search.h>
+#include <dirent.h>
+#include <string.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <errno.h>
+#include <libzfs.h>
+
+#define	STRCMP ((int(*)(const void *, const void *))&strcmp)
+#define	ABORT_ENOMEM() \
+	do { \
+		fprintf(stderr, PROGNAME ": %d: " \
+		    "not enough memory (L%d)!\n", getpid(), __LINE__); \
+		_exit(1 << 0); \
+	} while (0)
+
+#define	true ((_Bool)1)
+#define	false ((_Bool)0)
+
+
+#define	OUTPUT_HEADER \
+	"# Automatically generated by zfs-mount-generator\n" \
+	"\n"
+
+#define	PROGNAME "zfs-mount-generator"
+#define	FSLIST SYSCONFDIR "/zfs/zfs-list.cache"
+#define	ZFS SBINDIR "/zfs"
+
+/* Almost like the one in libzfs_util.c but not quite :) */
+#define	URI_REGEX_S "^\\([A-Za-z][A-Za-z0-9+.\\-]*\\):\\/\\/\\(.*\\)$"
+static regex_t uri_regex;
+
+static char *argv0;
+
+static const char *destdir = "/tmp";
+static int destdir_fd = -1;
+
+static void *known_pools = NULL; /* tsearch() of C strings */
+static struct {
+	sem_t noauto_not_on_sem;
```

Comment:
It would be preferable to use the pthreads interface instead of a semaphore.  We had one other instance of a semaphore we recently removed since it turns out they're not very portable, see d553fb9b9e18e04d1c85bf1930bcf0f46757d32e.  Of course, since systemd is Linux specific this probably would never be a issue but still it'd be nice to avoid things we know won't be portable.

Reply:
POSIX unnamed semaphores are a well-defined POSIX interface with the sole purpose of Doing Exactly This. As far as I can tell, there are no others (SysV semaphores also Exist, but the API is absolutely atrocious, and also not portable to the macintosh, since that's a BSD derivative). I doubt pthreads will work across processes (and expecting them to is, uh, lol; I'm certainly not in the market for testing that in perpetuity on all libcs past and future).

Reply:
Also, on a "macOS Catalina Version 10.15.7" (Darwin 19.6.0, Oct 2020), which is the latest version that can be virtualised, these do exist and function, so I get the supposed problem even less.

Reply:
My understanding was the functions still exist but were deprecated.  Either way since this is all going to be Linux-only code I'm fine with sticking with the semaphores. If it ever somehow becomes a problem we can deal with it as needed then.